src/Processors/Executors/PullingAsyncPipelineExecutor.cpp

```diff
@@ -179,10 +179,40 @@ void PullingAsyncPipelineExecutor::cancel()
         return;
 
     /// Cancel execution if it wasn't finished.
+    cancelWithExceptionHandling([&]()
+    {
+        data->executor->cancel();
+    });
+
+    /// The following code is needed to rethrow exception from PipelineExecutor.
+    /// It could have been thrown from pull(), but we will not likely call it again.
+
+    /// Join thread here to wait for possible exception.
+    if (data->thread.joinable())
+        data->thread.join();
+
+    /// Rethrow exception to not swallow it in destructor.
+    data->rethrowExceptionIfHas();
+}
+
+void PullingAsyncPipelineExecutor::cancelReading()
+{
+    if (!data)
+        return;
+
+    /// Stop reading from source if pipeline wasn't finished.
+    cancelWithExceptionHandling([&]()
+    {
+        data->executor->cancelReading();
+    });
+}
+
+void PullingAsyncPipelineExecutor::cancelWithExceptionHandling(CancelFunc && cancel_func)
+{
     try
     {
         if (!data->is_finished && data->executor)
```

Comment:
Move check into `cancel_func`. I think it's more clear, because I use `if (!data) return` check at the beginning of `cancel` functions. So all `data` checks can be seen in the same place.